Add dispatch keys for out-of-tree grad+vmap prototype

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56824 Add dispatch keys for out-of-tree grad+vmap prototype**

This PR adds 6 dispatch uses to be used with prototyping.

I'm not sure what the best way to name these are, please let me know if
you think that these should have the same prefix.

Test Plan:
- wait for tests

Differential Revision: [D27999963](https://our.internmc.facebook.com/intern/diff/D27999963)

zou3519 committed:
Add dispatch keys for out-of-tree grad+vmap prototype

This PR adds 6 dispatch uses to be used with prototyping.

I'm not sure what the best way to name these are, please let me know if
you think that these should have the same prefix.

Test Plan:
- wait for tests

[ghstack-poisoned]

Chillee reviewed (approved):
Python key looks good to me :)

albanD commented on c10/core/DispatchKey.cpp:
```diff
@@ -126,6 +126,25 @@ const char* toString(DispatchKey t) {
+    // https://github.com/zou3519/functorch
+    // We plan on eventually upstreaming the prototype into core, at which
+    // point it will have a different design that should use fewer keys.
+    case DispatchKey::Python:
```
nit: could this one be more specific? Or at least have a less generic name?
If this is specifically for FX or torch_function, maybe we want to mention it? Or even FuncTorchPython if this is specific to this lib.

zou3519 replied:
In the end state, we do want a key whose name is Python: https://github.com/pytorch/pytorch/issues/55095. The idea is that all `__torch_function__` objects should have the ability to go into the dispatcher using this key

albanD replied:
But that would be part of the "upstreaming the prototype into core, at which point it will have a different design" ?
In particular, it is simple to delete a key that is used just for prototype, it might be harder to re-purpose one without changing the name?
But if you think that this is going to be the final behavior, its ok with me.

zou3519 replied:
Worst comes to worst we can rename the key. But I do see the concern. What do you think about prefixing all of the keys with "Functorch" to avoid confusion?

albanD replied:
I think a FuncTorch prefix is good.

It also has the added benefit when you move to core and you have "proper" keys that match, you can simply alias the FuncTorch key to the corresponding core key and you're done :)

zou3519 replied:
Cool, will do!

albanD reviewed (approved):
Looks good to me.
The naming of the Python key feels a bit too generic if someone stumbles upon it outside of this context though.

zou3519 committed:
Update on "Add dispatch keys for out-of-tree grad+vmap prototype"


This PR adds 6 dispatch uses to be used with prototyping.

I'm not sure what the best way to name these are, please let me know if
you think that these should have the same prefix.

Test Plan:
- wait for tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add dispatch keys for out-of-tree grad+vmap prototype"


This PR adds 6 dispatch uses to be used with prototyping.

I'm not sure what the best way to name these are, please let me know if
you think that these should have the same prefix.

Test Plan:
- wait for tests

Differential Revision: [D27999963](https://our.internmc.facebook.com/intern/diff/D27999963)

[ghstack-poisoned]

ailzhang reviewed (approved):
LGTM!

zou3519 committed:
Update on "Add dispatch keys for out-of-tree grad+vmap prototype"


This PR adds 6 dispatch uses to be used with prototyping.

I'm not sure what the best way to name these are, please let me know if
you think that these should have the same prefix.

Test Plan:
- wait for tests

Differential Revision: [D27999963](https://our.internmc.facebook.com/intern/diff/D27999963)

[ghstack-poisoned]
